subprojects/functional/src/main/java/org/gradle/internal/Deferrable.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * An invocation to be executed at most once, but one that can be deferred.
+ *
+ * @param <T> The type which will be computed.
+ */
+public interface Deferrable<T> {
+
+    /**
+     * The result of the invocation when it is already available.
+     */
+    Optional<T> getCompleted();
+
+    /**
+     * Obtain the result of the invocation, either by returning the already computed result or by computing it synchronously.
+     *
+     * A result is only calculated once.
+     */
+    T completeAndGet();
+
+    /**
+     * Maps the result of the invocation via a mapper.
```

Comment:
I think we should declare what happens when the `mapper` returns null on the value. Looking a the implementation, it seems that the completed value then becomes absent because of the behavior of `Optional`. Is that intended? Or is the idea that the `mapper` must not return `null`?

Reply:
OK, I added some tests and ensured that when `mapper` returns `null` we throw.